src/tools/U8U16Test/main.cpp

```diff
@@ -0,0 +1,429 @@
+{
+    PrintHeader(__func__);
+    const size_t endLoop{ u8Str.length() / u16ChunkLen };
+    double duration{};
+    size_t length{};
+    HRESULT hRes{};
+    std::wstring u16Str{};
+
+    for (size_t i{}; i < endLoop; i += u8CharLen)
+    {
+        const std::string_view sv{ &u8Str.at(i), u16ChunkLen * u8CharLen };
+        GetDuration();
+        hRes = u8u16(sv, u16Str);
+        duration += GetDuration();
+        length += u16Str.length();
+    }
+
+    const wchar_t randElem16 = u16Str.at(RandomIndex(static_cast<ptrdiff_t>(u16Str.length())));
+    std::cout << " ignore me " << static_cast<int>(randElem16)
+              << "\n HRESULT " << hRes << "\n length " << length << "\n elapsed " << duration << std::endl;
+}
+
+void u8u16_ptr_Chunks(std::string_view u8Str, size_t u8CharLen, size_t u16ChunkLen)
+{
+    PrintHeader(__func__);
+    const size_t endLoop{ u8Str.length() / u16ChunkLen };
+    double duration{};
+    size_t length{};
+    HRESULT hRes{};
+    std::wstring u16Str{};
+
+    for (size_t i{}; i < endLoop; i += u8CharLen)
+    {
+        const std::string_view sv{ &u8Str.at(i), u16ChunkLen * u8CharLen };
+        GetDuration();
+        hRes = u8u16_ptr(sv, u16Str);
+        duration += GetDuration();
+        length += u16Str.length();
+    }
+
+    const wchar_t randElem16 = u16Str.at(RandomIndex(static_cast<ptrdiff_t>(u16Str.length())));
+    std::cout << " ignore me " << static_cast<int>(randElem16)
+              << "\n HRESULT " << hRes << "\n length " << length << "\n elapsed " << duration << std::endl;
+}
+
+int main()
+{
+    // UTF-16 string length
+    constexpr const size_t u16Length{ 100000000u }; // 100,000 code points
+    //constexpr const size_t u16Length{ 10000000u }; // 10,000 code points
+
+    // chunk length in code points
+    constexpr const size_t chunkLen = 10u;
+
+    // UTF-16 character to be used
+    //const std::wstring testU16(u16Length, static_cast<wchar_t>(0x007E)); // TILDE (1 Byte in UTF-8)
+    //const std::wstring testU16(u16Length, static_cast<wchar_t>(0x00F6)); // LATIN SMALL LETTER O WITH DIAERESIS (2 Bytes in UTF-8)
+    const std::wstring testU16(u16Length, static_cast<wchar_t>(0x20AC)); // // EURO SIGN (3 Bytes in UTF-8)
+
+    HMODULE ntdll = LoadLibraryA("ntdll.dll");
+    if (ntdll != nullptr)
+    {
+        p_RtlUTF8ToUnicodeN = reinterpret_cast<t_RtlUTF8ToUnicodeN>(GetProcAddress(ntdll, "RtlUTF8ToUnicodeN"));
+        p_RtlUnicodeToUTF8N = reinterpret_cast<t_RtlUnicodeToUTF8N>(GetProcAddress(ntdll, "RtlUnicodeToUTF8N"));
+        if (!p_RtlUTF8ToUnicodeN || !p_RtlUnicodeToUTF8N)
+        {
+            FreeLibrary(ntdll);
+            return 1;
+        }
+    }
+    else
+    {
+        return 1;
+    }
+
+    std::string u8Str{};
+
+    std::cout << "### UTF-16 To UTF-8 ###" << std::endl;
+
+    WideCharToMultiByte_WholeString(testU16);
+    RtlUnicodeToUTF8N_WholeString(testU16);
+    u16u8_WholeString(testU16, u8Str);
+    u16u8_ptr_WholeString(testU16, u8Str);
```

Comment:
This might technically be redundant, but should we maybe also link this with types.lib and run these tests with the `til` implementation as well, just to make sure the `til` copy of these conversions doesn't drift to far?

Reply:
I agree. So, I'll try to link the lib. Furthermore I will update the unit tests with comparisons of the time taken using MB2WC/WC2MB and using u8u16/u16u8, then let them only pass if the latter have a better performance. That way you will always know if updates of the code would still be faster.

Reply:
I'm afraid I don't know enough to link contypes.lib. And since current test show a rather comparable performance I'd rather skip to update the unit tests to compare the performance with the API functions is this is still not blocking.